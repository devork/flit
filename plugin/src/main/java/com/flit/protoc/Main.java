package com.flit.protoc;

import com.google.protobuf.compiler.PluginProtos;

/**
 * Main entry point to the plugin which handles the parsing of the input from the main protoc process
 * and hands off to a {@link Plugin} instance.
 */
public class Main {

  public static void main(String[] args) throws Exception {
    Plugin plugin = new Plugin(PluginProtos.CodeGeneratorRequest.newBuilder().mergeFrom(System.in).build());
    plugin.process().writeTo(System.out);
  }

}
