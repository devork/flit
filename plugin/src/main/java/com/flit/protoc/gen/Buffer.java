package com.flit.protoc.gen;

/**
 * Holds a generation buffer used to write basic strings to a wrapper string builder
 */
public class Buffer {

  private StringBuilder buffer;
  private Indent indent;

  public Buffer() {
    this.indent = new Indent();
    this.buffer = new StringBuilder();
  }

  // write
  public Buffer w(String s) {
    buffer.append(s);
    return this;
  }

  // write (multi)
  public Buffer w(String... ss) {

    for (String s : ss) {
      buffer.append(s);
    }
    return this;
  }

  // indent write
  public Buffer iw(String s) {
    buffer.append(indent).append(s);
    return this;
  }

  // indent write (multi)
  public Buffer iw(String... ss) {
    buffer.append(indent);

    for (String s : ss) {
      buffer.append(s);
    }
    return this;
  }

  // write newline
  public Buffer wn(String s) {
    buffer.append(s).append("\n");
    return this;
  }

  // write newline (multi)
  public Buffer wn(String... ss) {

    for (String s : ss) {
      buffer.append(s);
    }
    buffer.append("\n");
    return this;
  }

  // indent write newline
  public Buffer iwn(String s) {
    buffer.append(indent).append(s).append("\n");
    return this;
  }

  // indent write newline (multi)
  public Buffer iwn(String... ss) {
    buffer.append(indent);

    for (String s : ss) {
      buffer.append(s);
    }
    buffer.append("\n");
    return this;
  }

  // newline
  public Buffer n() {
    buffer.append("\n");
    return this;
  }

  // increase the indentation
  public Buffer inc() {
    indent.inc();
    return this;
  }

  // decrease the indentation
  public Buffer dec() {
    indent.dec();
    return this;
  }

  public String toString() {
    return this.buffer.toString();
  }

}
