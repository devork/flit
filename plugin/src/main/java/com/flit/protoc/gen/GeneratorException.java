package com.flit.protoc.gen;

public class GeneratorException extends RuntimeException {
  public GeneratorException(String message) {
    super(message);
  }
}
