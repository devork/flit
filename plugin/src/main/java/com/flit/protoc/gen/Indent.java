package com.flit.protoc.gen;

public class Indent {
  private final String indent;
  private String current = "";

  public Indent() {
    this.indent = "    ";
  }

  public Indent(String indent) {
    this.indent = indent;
  }

  public String inc() {
    current += indent;
    return current;
  }

  public String dec() {
    if (current.isEmpty()) {
      return current;
    }

    current = current.substring(indent.length());
    return current;
  }

  public String toString() {
    return current;
  }
}
