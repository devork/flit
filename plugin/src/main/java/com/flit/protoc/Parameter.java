package com.flit.protoc;

import com.flit.protoc.gen.GeneratorException;
import lombok.Getter;
import lombok.ToString;

import java.util.Arrays;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

@Getter @ToString public class Parameter {

  public static final String PARAM_TARGET = "target";
  public static final String PARAM_CLIENT = "client";
  public static final String PARAM_TYPE = "type";
  public static final String PARAM_CONTEXT = "context";

  private String key;
  private String value;

  public Parameter(String[] strings) {
    this.key = strings[0];
    this.value = strings[1];
  }

  public static Map<String, Parameter> of(String value) {
    if (value == null || (value = value.trim()).isEmpty()) {
      throw new GeneratorException("Empty value passed to parameter builder");
    }

    return Arrays.stream(value.split(","))
      .map(p -> p.split("=", 2))
      .filter(s -> s.length == 2 && !s[0].trim().isEmpty() && !s[1].trim().isEmpty())
      .map(Parameter::new)
      .collect(Collectors.toMap(Parameter::getKey, Function.identity()));
  }
}
