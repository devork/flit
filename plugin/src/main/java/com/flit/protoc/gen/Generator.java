package com.flit.protoc.gen;

import com.flit.protoc.Parameter;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorRequest;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorResponse;

import java.util.List;
import java.util.Map;

/**
 * Defines the handler for the generator step
 */
public interface Generator {

  /**
   * This is the main entry point to code generation: the implementation is decided in the {@link com.flit.protoc.Plugin}
   * class from the given parameters.
   *
   * @param request   The inbound protoc request
   * @param params    The plugin parameters
   *
   * @return The list of files to be added to the output.
   */
  List<CodeGeneratorResponse.File> generate(CodeGeneratorRequest request, Map<String, Parameter> params);
}
