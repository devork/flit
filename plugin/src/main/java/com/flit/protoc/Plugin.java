package com.flit.protoc;

import com.flit.protoc.gen.Generator;
import com.flit.protoc.gen.GeneratorException;
import com.flit.protoc.gen.server.spring.SpringGenerator;
import com.flit.protoc.gen.server.undertow.UndertowGenerator;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorRequest;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorResponse;

import java.util.Map;

import static com.flit.protoc.Parameter.PARAM_TARGET;
import static com.flit.protoc.Parameter.PARAM_TYPE;

public class Plugin {

  private final CodeGeneratorRequest request;

  public Plugin(CodeGeneratorRequest request) {
    this.request = request;
  }

  public CodeGeneratorResponse process() {
    if (!request.hasParameter()) {
      return CodeGeneratorResponse.newBuilder().setError("Usage: --flit_out=target=server,type=[spring|undertow]:<PATH>").build();
    }

    Map<String, Parameter> params = Parameter.of(request.getParameter());
    try {
      CodeGeneratorResponse.Builder builder = CodeGeneratorResponse.newBuilder();
      resolveGenerator(params).generate(request, params).forEach(builder::addFile);
      return builder.build();
    } catch (GeneratorException e) {
      return CodeGeneratorResponse.newBuilder().setError(e.getMessage()).build();
    }
  }

  private Generator resolveGenerator(Map<String, Parameter> params) {
    if (!params.containsKey(PARAM_TARGET)) {
      throw new GeneratorException("No argument specified for target");
    }
    if (!params.containsKey(PARAM_TYPE)) {
      throw new GeneratorException("No argument specified for type");
    }
    switch (params.get(PARAM_TARGET).getValue()) {
      case "server":
        switch (params.get(PARAM_TYPE).getValue()) {
          case "boot":
          case "spring":
            return new SpringGenerator();
          case "undertow":
            return new UndertowGenerator();
          default:
            throw new GeneratorException("Unknown server type: " + params.get(PARAM_TYPE).getValue());
        }
      default:
        throw new GeneratorException("Unknown target type: " + params.get(PARAM_TARGET).getValue());
    }
  }
}
