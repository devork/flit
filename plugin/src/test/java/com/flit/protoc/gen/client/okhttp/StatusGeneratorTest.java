package com.flit.protoc.gen.client.okhttp;

import com.flit.protoc.Plugin;
import com.flit.protoc.gen.BaseGeneratorTest;
import com.google.protobuf.compiler.PluginProtos;
import org.approvaltests.Approvals;
import org.junit.Test;

import static java.util.stream.Collectors.toList;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

/**
 * Tests the generation of a service that has core definition imported from another file
 */
public class StatusGeneratorTest extends BaseGeneratorTest {

  @Test public void test_Generate() throws Exception {
    PluginProtos.CodeGeneratorRequest request = loadJson("status.okhttp.json");

    Plugin plugin = new Plugin(request);
    PluginProtos.CodeGeneratorResponse response = plugin.process();

    assertNotNull(response);
    assertEquals(2, response.getFileCount());

    assertEquals(response.getFile(0).getName(), "com/example/helloworld/RpcStatus.java");
    assertEquals(response.getFile(1).getName(), "com/example/helloworld/RpcStatusClient.java");

    Approvals.verifyAll("", response.getFileList().stream().map(f -> f.getContent()).collect(toList()));
    response.getFileList().forEach(f -> assertParses(f));
  }

}
