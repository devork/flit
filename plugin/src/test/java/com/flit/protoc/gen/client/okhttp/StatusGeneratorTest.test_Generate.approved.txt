[0] = package com.example.helloworld;

public interface RpcStatus {
  StatusOuterClass.StatusResponse dispatchGetStatus(Core.Empty in);
}

[1] = package com.example.helloworld;

import com.flit.runtime.ErrorCode;
import com.flit.runtime.FlitException;
import java.io.InputStream;
import java.lang.Exception;
import java.lang.String;
import java.util.function.Function;
import okhttp3.HttpUrl;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class RpcStatusDispatcher {
  public static final String SERVICE_PATH_PREFIX = "/twirp/com.example.helloworld.Status";

  private final String baseAddress;

  private OkHttpClient client;

  Function<BeforeRequestPayload, Request.Builder> beforeRequest;

  public RpcStatusDispatcher(String address, OkHttpClient client) {
    this.baseAddress = address;
    this.client = client;
  }

  public void setBeforeRequest(Function<BeforeRequestPayload, Request.Builder> beforeRequest) {
    this.beforeRequest = beforeRequest;
  }

  public StatusOuterClass.StatusResponse dispatchGetStatus(Core.Empty in) throws Exception {
    HttpUrl url = HttpUrl.parse(baseAddress + SERVICE_PATH_PREFIX + "GetStatus");
    RequestBody requestBody = RequestBody.create(in.toByteArray(), MediaType.get("application/protobuf"));
    Request.Builder builder = new Request.Builder();
    if(beforeRequest != null) {
      builder = beforeRequest.apply(new BeforeRequestPayload(builder, url));
    }
    builder.addHeader("Accept", "application/protobuf");
    builder.addHeader("Content-Type", "application/protobuf");
    builder.addHeader("Flit-Version", "v1.1.0");
    builder.url(url);
    builder.post(requestBody);
    Request request = builder.build();
    String responseString;
    InputStream responseStream;
    try(Response response = client.newCall(request).execute()) {
      if(response.code() != 200) {
        responseString = response.body().toString();
        throw FlitException.builder().withErrorCode(ErrorCode.INTERNAL).withMeta("message", responseString).withMessage("RPC error").build();
      } else {
        responseStream = response.body().byteStream();
        return StatusOuterClass.StatusResponse.parseFrom(responseStream);
      }
    }
  }

  public class BeforeRequestPayload {
    public Request.Builder builder;

    public HttpUrl url;

    public BeforeRequestPayload(Request.Builder builder, HttpUrl url) {
      this.builder = builder;
      this.url = url;
    }
  }
}

